test/python/circuit/test_matrix_gate.py

```diff
@@ -0,0 +1,143 @@
+        self.assertIsInstance(dnode['op'], UnitaryMatrixGate)
+        for qubit in dnode['qargs']:
+            self.assertTrue(qubit[1] in [0, 1])
+        self.assertTrue(numpy.allclose(dnode['op'].representation,
+                                       matrix))
+
+    def test_2q_unitary(self):
+        backend = BasicAer.get_backend('qasm_simulator')
+        qr = QuantumRegister(2)
+        cr = ClassicalRegister(2)
+        qc = QuantumCircuit(qr, cr)
+        σx = numpy.array([[0, 1], [1, 0]])
+        σy = numpy.array([[0, -1j], [1j, 0]])
+        matrix = numpy.kron(σx, σy)
+        qc.x(qr[0])
+        qc.unitary(matrix, qr[0], qr[1])
+        pm = PassManager()
+        pm.append(CXCancellation())
+        qc2 = transpile(qc, backend, pass_manager=pm)
+        # test of qasm output
+        self.log.info(qc2.qasm())
+        # test of text drawer
+        self.log.info(qc2)
+        dag = circuit_to_dag(qc)
+        nodes = dag.twoQ_nodes()
+        self.assertTrue(len(nodes) == 1)
+        dnode = nodes[0]
+        self.assertIsInstance(dnode['op'], UnitaryMatrixGate)
+        for qubit in dnode['qargs']:
+            self.assertTrue(qubit[1] in [0, 1])
+        self.assertTrue(numpy.allclose(dnode['op'].representation,
+                                       matrix))
+
+    def test_3q_unitary(self):
+        backend = BasicAer.get_backend('qasm_simulator')
+        qr = QuantumRegister(4)
+        qc = QuantumCircuit(qr)
+        σx = numpy.array([[0, 1], [1, 0]])
+        σy = numpy.array([[0, -1j], [1j, 0]])
+        matrix = numpy.kron(σy, numpy.kron(σx, σy))
+        qc.x(qr[0])
+        qc.unitary(matrix, qr[0], qr[1], qr[3])
+        qc.cx(qr[3], qr[2])
+        # test of qasm output
+        self.log.info(qc.qasm())
+        # test of text drawer
+        self.log.info(qc)
+        dag = circuit_to_dag(qc)
+        nodes = dag.threeQ_or_more_nodes()
+        self.assertTrue(len(nodes) == 1)
+        dnode = nodes[0][1]
+        self.assertIsInstance(dnode['op'], UnitaryMatrixGate)
+        for qubit in dnode['qargs']:
+            self.assertTrue(qubit[1] in [0, 1, 3])
+        self.assertTrue(numpy.allclose(dnode['op'].representation,
+                                       matrix))
+
+    def test_qobj_with_unitary_matrix(self):
+        backend = BasicAer.get_backend('qasm_simulator')
+        qr = QuantumRegister(4)
+        qc = QuantumCircuit(qr)
+        σx = numpy.array([[0, 1], [1, 0]])
+        σy = numpy.array([[0, -1j], [1j, 0]])
+        matrix = numpy.kron(σy, numpy.kron(σx, σy))
+        qc.x(qr[0])
+        qc.unitary(matrix, qr[0], qr[1], qr[3])
+        qc.cx(qr[3], qr[2])
+        qobj = circuits_to_qobj(qc)
+        instr = qobj.experiments[0].instructions[1]
+        self.assertEqual(instr.name, 'unitary')
+        self.assertTrue(numpy.allclose(
+            numpy.array(instr.params).astype(numpy.complex64),
+            matrix))
+
+    def test_aer(self):
```

Comment:
let's move this as a test in Aer, under this folder: https://github.com/Qiskit/qiskit-aer/tree/master/test/terra

aer is not a dependency of terra, so this will fail for people who only have BasicAer.

Reply:
ok